### ISSUE TITLE
Problems with lambda encoding.

### ISSUE THREAD

**Issue by Kesanov** [ethereum/moon-lang#36]
```haskell
True  a _ = a
False _ a = a

not bool = bool False True
```

I see two problems with lambda encoded (scott in this case) data types.

**1. It requires laziness for constructors with zero arguments.**

`True [] [1..1e+20]` currently evaluates both the first and second branch, which is extremely inefficient.

**2. It does not allow [join](https://www.microsoft.com/en-us/research/publication/compiling-without-continuations/) related optimizations.**

It is not possible to fuse `b => not (not b)` into `b => b` because we do not know, what `b` does with its arguments.  On the other hand with Haskell ADT it is simple to optimize `not . not` into `id` because 

``` haskell
not . not

\b -> not (not b)

\b -> not $ case b of
  True -> False
  False -> True

\b -> case b of
  True -> not False
  False -> not True

\b -> case b of
  True -> True
  False -> False

\b -> b
```

**Comment by VictorTaelin** [comment]
Yep, (1) is a problem. I think lazy/force primitives as previously suggested would be a great solution. Alternatively, it wouldn't be hard to make the whole implementation lazy, although that'd certainly hit the performance considerably. A GHC-based evaluator wouldn't have that problem, obviously.

For (2), how so? `not . not` reduces to `id` fine. Mind this program:

```haskell
true = t => f => t
false = t => f => f
dot = f => g => x => (f (g x))
not = bool => t => f => (bool f t)
(dot not not)
```

This outputs:

```haskell
a => b => c => (a b c)
```

Which is the identity function (just not eta-reduced, as Moon doesn't eta-reduce).

**Comment by Kesanov** [comment]
The problem lies in `not` implementation:  `not = bool => (bool false true)`
```haskell
true = t => f => t
false = t => f => f
dot = f => g => x => (f (g x))
not = bool => (bool false true)
(dot not not)
```
outputs
```haskell
a => (a b => c => c b => c => b b => c => c b => c => b)
```

Unless you are the author of library, you want to use the exported constructors `false` and `true` (or `val` and `end` etc.) since you don't want to depend on internal implementation and that means you are screwed.

**Comment by Kesanov** [comment]
Maybe we do not want to export `false` and `true` but something like `getBoolConstructors = body => t => f => (body t f)`. 

Then `not = bool => (getBoolConstructors t => f => bool f t)` and `(dot not not)` outputs `a => b => c => (a b c)` correctly.

**Comment by VictorTaelin** [comment]
This makes sense actually, interesting... it plays well with the idea of having a generics system based on building ADT descriptions.

**Comment by Kesanov** [comment]
> it plays well with the idea of having a generics system based on building ADT description

What do you mean?
